inspirehep/modules/workflows/tasks/manual_merging.py

```diff
@@ -0,0 +1,165 @@
+from inspirehep.utils.record_getter import get_db_record, RecordGetterError
+from inspirehep.modules.workflows.utils import with_debug_logging
+from invenio_db import db
+
+
+def _check_sources(wf_obj, head, update):
+    # Many records in Inspire don't have an acquisition_source.
+    # If the records to merge also don't, they are treated as arXiv records.
+    # We expect the `manual_merge` workflow to work on arXiv-publisher or
+    # publisher-arXiv records. In other cases, the arXive-arXiv strategy
+    # will be used, enforcing acquisition_source field to arXiv.
+    source_head = wf_obj.extra_data.get('head_source')
+    source_updated = wf_obj.extra_data.get('update_source')
+
+    if head.get('acquisition_source') is None:
+        head['acquisition_source'] = {'source': source_head or 'arxiv'}
+
+    if update.get('acquisition_source') is None:
+        update['acquisition_source'] = {'source': source_updated or 'arxiv'}
+
+
+def _get_head_and_update(obj):
+    """
+    Retrieve from the database the records defined by their id in the
+    workflow object, in according to the pid_type value in `obj`.
+
+    Arguments:
+        obj(WorkflowObject): the current workflow object
+
+    Returns:
+        tuple(InspireRecord): the two records to merge
+    """
+    pid_type = obj.data.get('pid_type')
+    recid_head = obj.data.get('recid_head')
+    recid_update = obj.data.get('recid_update')
+
+    try:
+        head = get_db_record(pid_type, recid_head)
+        update = get_db_record(pid_type, recid_update)
+    except RecordGetterError as e:
+        raise ManualMergeError(
+            'No record found for PID_TYPE={} and PID={}'
+            .format(pid_type, str(e.message) or "''")
+        )
+    return head, update
+
+
+@with_debug_logging
+def merge_records(obj, eng):
+    """
+    Merge the records which ids are defined in the `obj` parameter and store
+    the merged record and relative conflicts in `obj.data['merged']` and
+    `obj.data['conflicts']`.
+    """
+    head, update = _get_head_and_update(obj)
+    _check_sources(obj, head, update)
+
+    merged, conflicts = inspire_json_merge(
+        {},
+        head,
+        update,
+        # TODO: pass pid_type when it will be able to merge other collections
+    )
+
+    obj.data['merged'] = merged
+    obj.data['conflicts'] = conflicts
+
+
+@with_debug_logging
+def ask_for_approval(obj, eng):
+    """Stop the Workflow engine"""
+    eng.halt('Manual Merge halted for curator approval.')
+
+
+@with_debug_logging
+def edit_metadata_and_store(obj, eng):
+    """
+    Replace the `head` record with the previously merged record and updates
+    some reference in order to delete the `update` record, linking it to the
+    new `head`.
+    """
+    head_rec, update_rec = _get_head_and_update(obj)
+
+    merged_rec = InspireRecord.create(obj.data['merged'])
+
+    head_rec.update(merged_rec)  # head's content will be replaced by merged
```

Comment:
Put before a `head_rec.clear()`. Otherwise you have stale keys if they are not in `merged_rec`.